// ng-drag-drop/__ivy_ngcc__/src/ng-drag-drop.module.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var core_1 = require("@angular/core");
var draggable_directive_1 = require("./directives/draggable.directive");
var droppable_directive_1 = require("./directives/droppable.directive");
var ng_drag_drop_service_1 = require("./services/ng-drag-drop.service");
var ɵngcc0 = require('@angular/core');
var ɵngcc1 = require('./directives/draggable.directive');
var ɵngcc2 = require('./directives/droppable.directive');
var NgDragDropModule = /** @class */ (function () {
    function NgDragDropModule() {
    }
    NgDragDropModule.forRoot = function () {
        return {
            ngModule: NgDragDropModule,
            providers: [ng_drag_drop_service_1.NgDragDropService]
        };
    };
NgDragDropModule.ɵfac = function NgDragDropModule_Factory(t) { return new (t || NgDragDropModule)(); };
NgDragDropModule.ɵmod = /*@__PURE__*/ ɵngcc0.ɵɵdefineNgModule({ type: NgDragDropModule });
NgDragDropModule.ɵinj = /*@__PURE__*/ ɵngcc0.ɵɵdefineInjector({});
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && ɵngcc0.ɵsetClassMetadata(NgDragDropModule, [{
        type: core_1.NgModule,
        args: [{
                imports: [],
                declarations: [
                    draggable_directive_1.Draggable,
                    droppable_directive_1.Droppable
                ],
                exports: [
                    draggable_directive_1.Draggable,
                    droppable_directive_1.Droppable
                ]
            }]
    }], function () { return []; }, null); })();
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && ɵngcc0.ɵɵsetNgModuleScope(NgDragDropModule, { declarations: [ɵngcc1.Draggable, ɵngcc2.Droppable], exports: [ɵngcc1.Draggable, ɵngcc2.Droppable] }); })();
    return NgDragDropModule;
}());
exports.NgDragDropModule = NgDragDropModule;
